sharkfuser/include/fusilli/support/asm_emitter.h

```diff
@@ -554,6 +554,256 @@ inline std::string ConvFPropNode::emitNodePreAsm() const {
+
+  // Emit the permute op itself.
+  constexpr std::string_view schema = R"(
+    %none_DW_{0} = torch.constant.none
+    %dtype_DW_{0} = torch.constant.int {3}
+    %empty_{0} = torch.aten.empty.memory_format {1}, %dtype_DW_{0}, %none_DW_{0}, %none_DW_{0}, %none_DW_{0}, %none_DW_{0} : !torch.list<int>, !torch.int, !torch.none, !torch.none, !torch.none, !torch.none -> {2}
+  )";
+
+  torch_upstream::ScalarType dataType =
+      DataTypeToTorchType.at(DW->getDataType());
+  std::string output =
+      std::format(schema,
+                  suffix,                      // {0}
+                  "%" + prefix + "_" + suffix, // {1}
+                  DW->getTensorTypeAsm(/*isValueTensor=*/true,
+                                       /*useLogicalDims=*/true), // {2}
+                  std::to_string(static_cast<int>(dataType))     // {3}
+      );
+
+  return oss.str() + output;
+}
+
+inline std::string ConvWGradNode::emitNodePreAsm() const {
+  constexpr std::string_view schema = R"(
+    %bias_{0} = torch.constant.none
+    %transposed_{0} = torch.constant.bool false
+    %output_padding_{0} = torch.prim.ListConstruct  : () -> !torch.list<int>
+    %groups_{0} = torch.constant.int 1
+    {1}
+    {2}
+    {3}
+    {4}
+    {5}
+    {11}
+    %true_{0} = torch.constant.bool true
+    %false_{0} = torch.constant.bool false
+    %output_mask_{0} = torch.prim.ListConstruct %false_{0}, %true_{0}, %false_{0} : (!torch.bool, !torch.bool, !torch.bool) -> !torch.list<bool>
+    %grad_input_{0}, {6}_perm, %grad_bias_{0} = torch.aten.convolution_backward {7}, %empty_{0}, %bias_{0}, %stride_{0}, %padding_{0}, %dilation_{0}, %transposed_{0}, %output_padding_{0}, %groups_{0}, %output_mask_{0} : {8}, {12}, !torch.none, !torch.list<int>, !torch.list<int>, !torch.list<int>, !torch.bool, !torch.list<int>, !torch.int, !torch.list<bool> -> !torch.none, {9}, !torch.none
+    {10}
+    )";
+
+  // Suffix the SSA names of internal values (constant attributes) using
+  // the unique ConvWGradAttr name to avoid re-definition of names across
+  // the overall MLIR assembly.
+  std::string uniqueSSASuffix = convWGradAttr.getName();
+
+  std::string output =
+      std::format(schema,
+                  uniqueSSASuffix,            // {0}
+                  getStrideOpsAsm(),          // {1}
+                  getPaddingOpsAsm(),         // {2}
+                  getDilationOpsAsm(),        // {3}
+                  getPermuteDYOpsAsm(),       // {4}
+                  getPermuteXOpsAsm(),        // {5}
+                  getResultNamesAsm(),        // {6}
+                  getOperandNamesAsm(),       // {7}
+                  getOperandTypesAsm(),       // {8}
+                  getResultTypesAsm(),        // {9}
+                  getPermuteDWOpsAsm(),       // {10}
+                  getPermutedEmptyDWOpsAsm(), // {11}
+                  convWGradAttr.getDW()->getTensorTypeAsm(
+                      /*isValueTensor=*/true, /*useLogicalDims=*/true) // {12}
```

Comment:
Why is this separate? Should it just be `getResultTypesAsm()` here? (i.e. we don't need {12})